src/context.rs

```diff
@@ -1237,52 +1265,170 @@ impl SeafowlContext for DefaultSeafowlContext {
+                                    Err(error) => warn!(
+                                        "Failed constructing pruning statistics for table {} (version: {}) during UPDATE execution: {}",
+                                        table.name, table.table_version_id, error
+                                    )
+                                }
+                            }
+
+                            // Create a complete projection expression, including replicating
+                            // the columns omitted in the provided assignments
+                            let projection_exprs = schema
+                                .fields
+                                .iter()
+                                .map(|f| {
+                                    if assignments.contains_key(f.name()) {
+                                        let mut expr = create_physical_expr(
+                                            &assignments[f.name()],
+                                            &df_schema,
+                                            &schema,
+                                            &ExecutionProps::new(),
+                                        )?;
+
+                                        let data_type = f.data_type().clone();
+                                        if expr.data_type(&schema)? != data_type {
+                                            // Literal value potentially mistyped; try to re-cast it
+                                            expr = cast(expr, &schema, data_type)?;
+                                        }
+
+                                        // If the selection was specified, use a CASE WHEN
+                                        // (selection expr) THEN (assignment expr) ELSE old row
+                                        // approach
+                                        if let Some(sel_expr) = &selection_expr {
+                                            expr = case(
+                                                None,
+                                                vec![(sel_expr.clone(), expr)],
+                                                Some(col(f.name(), &schema)?),
+                                                &schema,
+                                            )?;
+                                        }
+
+                                        Ok((expr, f.name().to_string()))
+                                    } else {
+                                        // Just replicate the omitted columns
+                                        Ok((
+                                            col(f.name(), &schema)?,
+                                            f.name().to_string(),
+                                        ))
+                                    }
+                                })
+                                .collect::<Result<Vec<(Arc<dyn PhysicalExpr>, String)>>>(
+                                )?;
```

Comment:
I think this `assignments -> projection exprs` function is self-contained enough that it could be factored out (and potentially unit tested) to keep this UPDATE handler shorter

Reply:
Good point, extracted, though note that unti testing is a bit of a pain due to need to compare trait objects for equality.